sdk/dotnet/Pulumi/Testing/MockMonitor.cs

```diff
@@ -0,0 +1,123 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.WellKnownTypes;
+using Pulumi.Serialization;
+using Pulumirpc;
+
+namespace Pulumi.Testing
+{
+    internal class MockMonitor : IMonitor
+    {
+        private readonly IMocks _mocks;
+        private readonly Serializer _serializer = new Serializer();
+        private string? _rootResourceUrn = null;
+        
+        public readonly List<string> Errors = new List<string>();
+        public readonly List<Resource> Resources = new List<Resource>();
+
+        public ImmutableDictionary<string, object> StackOutputs { get; private set; } =
+            ImmutableDictionary<string, object>.Empty;
+
+        public MockMonitor(IMocks mocks)
+        {
+            _mocks = mocks;
+        }
+
+        public async Task<InvokeResponse> InvokeAsync(InvokeRequest request)
+        {
+            var result = await _mocks.CallAsync(request.Tok, ToDictionary(request.Args), request.Provider);
+            return new InvokeResponse { Return = await SerializeAsync(result) };
+        }
+
+        public async Task<ReadResourceResponse> ReadResourceAsync(Resource resource, ReadResourceRequest request)
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name,
+                ToDictionary(request.Properties), request.Provider, request.Id);
+            this.Resources.Add(resource);
```

Comment:
this needs to be threadsafe.  (applies to other functions as well).

Reply:
added a lock around it